pages/developers/nox.md

```diff
@@ -0,0 +1,255 @@
+permanently (`pipx install nox`) or by running `pipx run nox` instead of `nox`.
+
+
+On GitHub Actions or Azure, pipx is available by default, so you should use
+`pipx run nox`. To give it access to all Python versions, you can use this action:
+
+```yaml
+- uses: excitedleigh/setup-nox@v2.0.0
+```
+
+You can now access all versions of Python from nox. At least in GitHub Actions,
+you should add `--force-color` to your nox runs to get color output in your
+logs.
+
+### Introduction
+
+Nox is a tool for running tasks, called "sessions", inside temporary virtual
+environments. It is configured through Python and is designed to resemble
+PyTest. The file it looks for is called `noxfile.py` by default. This is an example of a simple nox file:
+
+```python
+import nox
+
+@nox.session
+def tests(session: nox.Session) -> None:
+    """
+    Run the unit and regular tests.
+    """
+    session.install(".[test]")
+    session.run("pytest", *session.posargs)
+```
+
+This will create a session called `tests`. The function receives the "session"
+argument, which gives you access to the virtual environment it creates. You can
+use `.install()` to install inside the environment, and `.run()` to run inside
+the environment. We are also using `session.posargs` to allow extra arguments
+to be passed through to pytest. There are
+[more useful methods](https://nox.thea.codes/en/stable/config.html#module-nox.sessions)
+as well. 
+
+You can run this using:
+
+```console
+$ nox -s tests
+```
+
+You can see all defined sessions (along with the docstrings) using:
+
+```console
+$ nox -l
+```
+
+It is a good idea to list the sessions you want by default by setting `nox.options.sessions` near the top of your file:
+
+```python
+nox.options.sessions = ["lint", "tests"]
+```
+
+This will keep you from running extra things like `docs` by default.
+
+### Parametrizing
+
+You can parametrize sessions. either on Python or on any other item.
+
+```python
+# Shortcut to parametrize Python
+@nox.session(python=["3.6", "3.7])
+
+# General parametrization
+@nox.session
+@nox.parametrize("letter", ["a", "b"], ids=["a", "b"])
+def my_session(session: nox.Session, letter: str) -> None: ...
+```
+
+The optional `ids=` parameter can give the parametrization nice names, like in
+PyTest.
+
+If a user does not have a particular version of Python installed, it will be
+skipped. You can use a Docker container to run in an environment where all
+Python's (3.6+) are available:
+
+```console
+$ docker run --rm -itv $PWD:/src -w /src quay.io/pypa/manylinux_2_24_x86_64:latest pipx run nox
+```
+
+Another container you can use is `thekevjames/nox:latest`; this has nox
+pre-installed (no pipx) and Python 2.7 and 3.5 as well.
+
+### Useful sessions
+
+Things like bumping the versions can be made sessions - since nox handles the
+environment for you, you can use any Python dependencies you like, and not have
+to worry about installing anything. Here are some commonly useful sessions that
```

Comment:
> not have to worry about installing anything

So the idea here is that this is easier then just telling people who want to develop to install all the dependencies ala

```console
python -m pip install -e .[all]
```

or something similar?

Reply:
If a person wants to contribute to your project, instead of assign them to install pre-commit, run pre-commit, create a virtual environment, enter the virtual environment, install `-s .[tests]`, and run `pytest`, you can just ask them to run `nox`. Assuming it is set to run `["lint", "tests"]`, they will get linting and testing immediately, for free, with no setup.

Reply:
If they start regularly committing, then they can set up those things. But it's easier if they don't have to just to get started. You can even select conda instead of virtualenv, parametrize over Python versions, etc.

Reply:
Okay so just lowering the barrier to entry for development as the primary goal. I can get behind that given the frustrations I've had with getting Bazel setup correctly before for TFP. :+1: